From dde3d7d055b844722bf845268a4082ed7e0905ab Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Fri, 23 Aug 2019 09:52:15 +0800
Subject: [PATCH] staging: bcm2835-camera: replace GFP_KERNEL with GFP_ATOMIC

With irq and preemption disabled status, allocate memory with "GFP_ATOMIC" should be
used. Otherwise it might cause system sleeping.

To avoid below call trace:

BUG: sleeping function called from invalid context at mm/slab.h:422
in_atomic(): 1, irqs_disabled(): 0, pid: 415, name: v4l2src0:src
Preemption disabled at:
[<ffffff80010bc368>] get_msg_context+0x48/0xc8 [bcm2835_v4l2]
CPU: 0 PID: 415 Comm: v4l2src0:src Tainted: G         C
4.18.41-yocto-standard #1
Hardware name: Raspberry Pi 3 Model B+ (DT)
Call trace:
  dump_backtrace+0x0/0x170
  show_stack+0x24/0x30
  dump_stack+0x80/0xa4
  ___might_sleep+0x14c/0x170
  __might_sleep+0x58/0x90
  kmem_cache_alloc+0x278/0x2f8
  radix_tree_node_alloc.constprop.13+0x50/0x108
  radix_tree_extend+0xb4/0x1b8
  idr_get_free+0x150/0x2b0
  idr_alloc_u32+0x74/0x108
  idr_alloc+0x44/0x80
  get_msg_context+0x64/0xc8 [bcm2835_v4l2]
  mmal_vchi_buffer_init+0x24/0x48 [bcm2835_v4l2]
  buffer_init+0x88/0x98 [bcm2835_v4l2]
  __vb2_queue_alloc+0x2f0/0x428 [videobuf2_common]
  vb2_core_reqbufs+0x138/0x3f0 [videobuf2_common]
  vb2_ioctl_reqbufs+0x78/0xa8 [videobuf2_v4l2]
  v4l_reqbufs+0x50/0x60 [videodev]
  __video_do_ioctl+0x138/0x330 [videodev]
  video_usercopy+0x294/0x678 [videodev]
  video_ioctl2+0x3c/0x50 [videodev]
  v4l2_ioctl+0x64/0x90 [videodev]
  do_vfs_ioctl+0xc4/0x910
  ksys_ioctl+0x84/0xb8
  sys_ioctl+0x34/0x48
  el0_svc_naked+0x30/0x34

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/staging/vc04_services/bcm2835-camera/mmal-vchiq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/vc04_services/bcm2835-camera/mmal-vchiq.c b/drivers/staging/vc04_services/bcm2835-camera/mmal-vchiq.c
index 51e5b04ff0f5..df4fbb0adce9 100644
--- a/drivers/staging/vc04_services/bcm2835-camera/mmal-vchiq.c
+++ b/drivers/staging/vc04_services/bcm2835-camera/mmal-vchiq.c
@@ -187,7 +187,7 @@ get_msg_context(struct vchiq_mmal_instance *instance)
 	 */
 	spin_lock(&instance->context_map_lock);
 	handle = idr_alloc(&instance->context_map, msg_context,
-			   0, 0, GFP_KERNEL);
+			   0, 0, GFP_ATOMIC);
 	spin_unlock(&instance->context_map_lock);
 
 	if (handle < 0) {
-- 
2.17.1

